### PR TITLE
Change synthetic source logic for constant_keyword

### LINKED ISSUE #117083: Synthetic source implementation for `constant_keyword` is inconsistent with stored source
Currently synthetic source for `constant_keyword` field is implemented to always write the constant value in `_source`. Most of the time such a value was not actually provided in the document (there is no reason to do that, it's constant) and therefore would not be in stored `_source`. This is problematic when using `constant_keyword` fields as a metadata fields that can be different between indices (f.e. reindexing an index into an index that has a different value of such a field). This scenario works with stored source since the values are never there but fails with synthetic source.

### PR DESCRIPTION
**Pull request by lkts**
See details in #117083.

Closes #117083.

### CODE REVIEW

**Review by kkrik-es — APPROVED** [review]
Consider adding coverage to the reindex test, in a follow-up.

File: x-pack/plugin/mapper-constant-keyword/src/test/java/org/elasticsearch/xpack/constantkeyword/mapper/ConstantKeywordFieldMapperTests.java [hunk 1]
```diff
@@ -333,6 +333,17 @@ public void testNullValueSyntheticSource() throws IOException {
         assertThat(syntheticSource(mapper, b -> {}), equalTo("{}"));
     }
 
+    public void testNoValueInDocumentSyntheticSource() throws IOException {
```
Comment (martijnvg) [inline x-pack/plugin/mapper-constant-keyword/src/test/java/org/elasticsearch/xpack/constantkeyword/mapper/ConstantKeywordFieldMapperTests.java:336]: Maybe in a follow up we can extend `DataGenerationHelper` to use constant keyword? That we would test this with reindex too in `ReindexChallengeRestIT` subclasses.

**Review by martijnvg — APPROVED** [review]
👍 LGTM

**Comment by lkts** [comment]
I don't think serverless failures are related.

**Comment by lkts** [comment]
I'll backport to 8.17 and 8.x when the branch is there.

**Comment by lkts** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[](https://github.com/elastic/elasticsearch/pull/117209)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

**Comment by lkts** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.16|[](https://github.com/elastic/elasticsearch/pull/117210)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

**Comment by lkts** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.15|[](https://github.com/elastic/elasticsearch/pull/117211)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

**Comment by lkts** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.17|[](https://github.com/elastic/elasticsearch/pull/117288)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

**Pull request merged**